lib/shared/sdk/usbboot/index.js

```diff
@@ -0,0 +1,318 @@
+  /* eslint-disable lodash/prefer-lodash-method */
+  return usb.listDevices().filter(isUsbBootCapableUSBDevice).map((device) => {
+  /* eslint-enable lodash/prefer-lodash-method */
+    device.name = `USB device (${device.deviceDescriptor.idVendor.toString(16)}:${device.deviceDescriptor.idProduct.toString(16)})`
+    return device
+  })
+}
+
+/**
+ * @summary Wait for a USB device
+ * @function
+ * @private
+ *
+ * @description
+ * This function returns the *first* device that matches
+ * the specified criteria.
+ *
+ * @param {Object} properties - properties of the expected device
+ * @param {Object} options - options
+ * @param {Number} options.retries - number of retries before giving up
+ * @param {Number} options.delay - milliseconds before each retry
+ * @fulfil {(Object|Undefined)} - device
+ * @returns {Promise}
+ *
+ * @example
+ * return waitForDevice({
+ *   deviceDescriptor: {
+ *     idVendor: 0x0a5c
+ *   }
+ * }, {
+ *   retries: 10,
+ *   delay: 500
+ * }).then((device) => {
+ *   if (device) {
+ *     console.log('Found!')
+ *   }
+ * })
+ */
+const waitForDevice = (properties, options) => {
+  return exports.scan().then((devices) => {
+    return _.find(devices, (device) => {
+      return _.matches(properties, device)
+    })
+  }).then((device) => {
+    if (_.isNil(device) && options.retries) {
+      return Bluebird.delay(options.delay).then(() => {
+        const INTERVAL_DECREASE_FACTOR = 1
+        return waitForDevice(properties, {
+          retries: options.retries - INTERVAL_DECREASE_FACTOR,
+          delay: options.delay
+        })
+      })
+    }
+
+    return device
+  })
+}
+
+const closeDevice = (device) => {
+  debug(`Closing device: ${device.name}`)
+  device.close()
+}
+
+const writeBootCode = (device, endpoint, bootCodeBuffer) => {
+  debug('Writing bootcode')
+  debug(`Bootcode buffer length: ${bootCodeBuffer.length}`)
+  const bootMessageBuffer = protocol.createBootMessageBuffer(bootCodeBuffer.length)
+
+  debug('Writing boot message buffer to out endpoint')
+  return protocol.write(device, endpoint, bootMessageBuffer).then(() => {
+    debug('Writing boot code buffer to out endpoint')
+    return protocol.write(device, endpoint, bootCodeBuffer)
+  }).then(() => {
+    debug('Reading return code from device')
+    return protocol.read(device, protocol.RETURN_CODE_LENGTH)
+  }).then((data) => {
+    const returnCode = data.readInt32LE()
+    debug(`Received return code: ${returnCode}`)
+
+    if (returnCode !== protocol.RETURN_CODE_SUCCESS) {
+      throw new Error(`Couldn't write the bootcode, got return code ${returnCode} from device`)
+    }
+  })
+}
+
+const startFileServer = (device, endpoint, files) => {
+  debug('Listening for file messages')
+  return protocol
+    .read(device, protocol.FILE_MESSAGE_LENGTH)
+    .then(protocol.parseFileMessageBuffer)
+
+    // We get this error message when reading a command
+    // from the device when the communication has ended
+    .catch({
```

Comment:
It could, however we'd need to return a buffer from the `.catch` block that `parseFileMessageBuffer` can parse, which is a bit inconvenient. By putting it afterwards, we can send the already parsed object.